src/error.rs

```diff
@@ -2,327 +2,239 @@
-	/// Check if the error is an I/O error indicating that the connection was aborted by the remote peer.
-	pub fn is_connection_aborted(&self) -> bool {
-		if let Self::Io(e) = &self {
-			e.kind() == std::io::ErrorKind::ConnectionAborted
+	/// Get this error as remote error message.
+	///
+	/// See [`Self::remote_error()`] for more details on what a remote error is.
+	pub fn as_remote_error(&self) -> Option<&str> {
+		if let private::InnerError::RemoteError(msg) = &self.inner {
+			Some(msg)
 		} else {
-			false
+			None
 		}
 	}
-}
-
-/// An error occurred while sending an update or stream message.
-#[derive(Debug, Error)]
-#[error("{0}")]
-pub enum SendMessageError {
-	/// An I/O error occurred.
-	Io(#[from] std::io::Error),
 
-	/// The payload of the message is too large to send.
-	PayloadTooLarge(#[from] PayloadTooLarge),
-
-	/// The request has already been closed.
-	RequestClosed(#[from] RequestClosed),
-
-	/// Failed to encode the message.
-	EncodeBody(Box<dyn std::error::Error + Send>),
-}
-
-impl SendMessageError {
-	/// Check if the error is an I/O error indicating that the connection was aborted by the remote peer.
-	pub fn is_connection_aborted(&self) -> bool {
-		if let Self::Io(e) = &self {
-			e.kind() == std::io::ErrorKind::ConnectionAborted
+	/// Get this error as remote error message.
+	///
+	/// See [`Self::remote_error()`] for more details on what a remote error is.
+	pub fn into_remote_error(self) -> Option<String> {
+		if let private::InnerError::RemoteError(msg) = self.inner {
+			Some(msg)
 		} else {
-			false
+			None
 		}
 	}
 }
 
-/// An error occurred while performing a service call.
-#[derive(Debug, Error)]
-#[error("{0}")]
-pub enum ServiceCallError {
-	/// Failed to encode or send the request.
-	SendRequest(#[from] SendRequestError),
-
-	/// Failed to receive or decode the response.
-	RecvResponse(#[from] RecvMessageError),
+impl From<std::io::Error> for Error {
+	fn from(other: std::io::Error) -> Self {
+		Self::io_error(other)
+	}
 }
 
-/// An error decoding a message.
-#[derive(Debug, Error)]
-#[error("{0}")]
-pub enum FromMessageError {
-	/// The service ID was not recognized.
-	UnexpectedServiceId(#[from] crate::error::UnexpectedServiceId),
+pub(crate) mod private {
+	use super::*;
 
-	/// Failed to decode the message body.
-	DecodeBody(Box<dyn std::error::Error + Send>),
-}
+	pub(crate) fn connection_aborted() -> Error {
+		InnerError::from(std::io::Error::from(std::io::ErrorKind::ConnectionAborted)).into()
+	}
 
-// Allow a ReadMessageError to be converted to a RecvMessageError automatically.
-impl From<ReadMessageError> for RecvMessageError {
-	fn from(other: ReadMessageError) -> Self {
-		match other {
-			ReadMessageError::Io(e) => e.into(),
-			ReadMessageError::MessageTooShort(e) => e.into(),
-			ReadMessageError::InvalidMessageType(e) => e.into(),
-			ReadMessageError::PayloadTooLarge(e) => e.into(),
-		}
+	#[derive(Debug, Error)]
+	#[error("{0}")]
+	#[doc(hidden)]
+	pub enum InnerError {
```

Comment:
This could also be done by implementing a From for all the error types within the enum right? 
impl From <..> for InnerError?

Reply:
What exactly do you mean? I've removed all specific error types except for UnexpectedMessageType (because it has special formatting).

Reply:
I mean if you have something like impl From <..> for InnerError for every error type within an enum, you achieve the same abstraction as what you have now right? All you do then is to propagate InnerError to the client and conversion to individual errors is done by the From functionality. Atleast I thought that was the use of From.

Reply:
`From` would allow implicit conversion with the try operator (question mark), but I removed almost all error types and made everything apart from `Error` private. There's nothing left to convert from.